File: Weather.java
/***************************************************************
 * Attire Decider Weather Analyzer
 * By: Keith Gladstone (keithag@princeton.edu)
 * Created in June 2015
 * 
 * This program takes advantage of the archiving ability of the
 * scraper. Namely, it can pull scraped data from a stored file
 * with a timestamp before deciding to pull data again from the Internet.
 * This method conserves API calls and reduces traffic, allowing for
 * rapid testing at no risk of disturbing the data source.
 * 
 * Requires Java Version 8
 * Assume that if input file exists, it is formatted correctly.
 *
 * Sample zip codes to test with 
 * 
 * Princeton : 08540    // for testing normal
 * Seattle   : 98101    // for testing rain
 * Miami     : 33101    // for testing heat
 * Chicago   : 60290    // for testing cool
 * Caribou, ME : 04736  // for testing cold
 * Lahaina, HI : 96761  // for testing different time of day
 * Schenectady, NY : 12345 // for interesting zip code
 ***************************************************************/
import java.time.LocalDateTime;
import java.io.File;
public class Weather
{
    /***************************************************************/
    /* What to Wear: given weather data output what to wear        */
    /* Enter any valid U.S. zip code as the first argument         */
    /***************************************************************/
    public static void main(String[] args)
    {
        final int CHRONINTERVAL = 30; // tolerance for age of input data, in minutes

        /*
         * Pre-determined user temperature preferences
         */
        final int HOT = 70;
        final int WARM = 60;
        final int COOL = 50;
        final int COLD = 35;
        final int FREEZING = 15;
        
        LocalDateTime now = LocalDateTime.now(); // grab current time
        
        /*
         * Handle commandline arguments
         */

        String zip = "08540";
        if (args != null && args.length > 0)
            zip = args[0]; // set zip code to argument on run
        String filename = "data/" + zip;

        /***************************************************************/

        /*
         * Read weather data from saved file
         */
        File f = new File(filename);
        String xmlText;

        if(!f.exists()) // Need new data because input file does not exist
        { 
            //StdOut.println("File didn't exist");
            WeatherScrape scraper = new WeatherScrape(zip);
        }

        In in = new In(filename);
        xmlText = in.readAll(); // GET DATA FROM EXISTING FILE

        // Capture the timestamp and print it
        String thenStr = getTime(xmlText);
        LocalDateTime then = LocalDateTime.parse(thenStr);
        //StdOut.println("That timestamp was: " + then);

        String thatZip = getZip(xmlText); // Parse zip code of input data


        /*
        * Refresh the weather data if necessary (mismatched location OR data too old)
        */
        if (!thatZip.equals(zip) || now.minusMinutes(CHRONINTERVAL).compareTo(then) > 0) // if at least XXX time later
        {
            WeatherScrape scraper = new WeatherScrape(zip);
        } 
        
        /***************************************************************/

        /*
         * Get HTML Data from input
         */
        String location = getTown(xmlText) + ", " + getState(xmlText); // UPDATE location
        double temp = Double.parseDouble(getTemperature(xmlText));
        String sky = getSky(xmlText);
        
        /*
         * Deal with rain variable
         */
        Double rainVal = Double.parseDouble(getRain(xmlText));
        boolean rain;
        if (rainVal == 0)
            rain = false;
        else
            rain = true;
       
        /*
         * Output collected HTML data
         */
        StdOut.println("Weather for " + location);
        StdOut.println("Current Temperature is: " + temp + " degrees Fahrenheit");
        StdOut.println("Sky is " + sky);
        StdOut.println("Rain is " + rainVal + " in.");
        StdOut.print("Consider wearing: ");
        
        
        /*
         * Output temperature analysis
         * This section needs to be catered to user preferences in the next version
         */

        tempAnalysis(temp, HOT, WARM, COOL, COLD, FREEZING);

        if (rain)
            StdOut.println("Bring an umbrella");
        if (sky.equals("Clear"))
            StdOut.println("Bring sunglasses");
    }
    


    /*-----------------------------------------------------------------------*/
    /* Analysis methods                                                      */
    /*-----------------------------------------------------------------------*/

    public static String getTime(String xmlText)
    {
        int start = xmlText.indexOf("<time>") + 6;
        int end = xmlText.indexOf("</time>");
        return xmlText.substring(start, end);
    }
    

    /***************************************************************/
    /* Extract sky from HTML meta-data                             */
    /***************************************************************/
    public static String getSky(String xmlText)
    {
        int start = xmlText.indexOf("<sky>") + 5;
        int end = xmlText.indexOf("</sky>");
        return xmlText.substring(start, end);
    }
    
    /***************************************************************/
    /* Extract town from HTML meta-data                            */
    /***************************************************************/
    public static String getTown(String xmlText)
    {
        int start = xmlText.indexOf("<town>") + 6;
        int end = xmlText.indexOf("</town>");
        return xmlText.substring(start, end);
    }

    /***************************************************************/
    /* Extract state from HTML meta-data                           */
    /***************************************************************/
    public static String getState(String xmlText)
    {
        int start = xmlText.indexOf("<state>") + 7;
        int end = xmlText.indexOf("</state>");
        return xmlText.substring(start, end);
    }

    /***************************************************************/
    /* Extract location from HTML meta-data                        */
    /***************************************************************/
    public static String getZip(String xmlText)
    {
        int start = xmlText.indexOf("<zip>") + 5;
        int end = xmlText.indexOf("</zip>");
        return xmlText.substring(start, end);
    }
    
    /***************************************************************/
    /* Extract temperature from HTML meta-data                     */
    /***************************************************************/
     public static String getTemperature(String xmlText)
    {
        int start = xmlText.indexOf("<temp>") + 6;
        int end = xmlText.indexOf("</temp>");
        return xmlText.substring(start, end);
    }
    
    /***************************************************************/
    /* Extract rain from HTML meta-data                            */
    /***************************************************************/
     public static String getRain(String xmlText)
    {
        int start = xmlText.indexOf("<rain>") + 6;
        int end = xmlText.indexOf("</rain>");
        return xmlText.substring(start, end);
    }
    
    /***************************************************************/
    /* Handle analysis of temperature given inputs                 */
    /***************************************************************/
    public static void tempAnalysis(double temp, int HOT, int WARM, int COOL, int COLD, int FREEZING)
    {
        if (temp >= HOT)
            StdOut.println("T-Shirt, Shorts");
        
        else if (temp <= HOT && temp >= WARM)
            StdOut.println("T-Shirt and Shorts + Layer");
        
        else if (temp <= WARM && temp >= COOL)
            StdOut.println("Long Pants, Light Jacket");
        
        else if (temp <= COOL && temp >= COLD)
            StdOut.println("Long Pants, Outer Layer and/or Light Jacket");
        
        else if (temp <= COLD && temp >= FREEZING)
            StdOut.println("Long Pants, Winter Jacket, Hat");
        
        else if (temp <= FREEZING)
            StdOut.println("FREEZING: minimize outdoor exposure");
    }
}

File: Weather.py
# /***************************************************************
#  * Attire Decider Weather Analyzer <PYTHON VERSION>
#  * By: Keith Gladstone (keithag@princeton.edu)
#  * Created in June 2015
#  * 
#  * This program takes advantage of the archiving ability of the
#  * scraper. Namely, it can pull scraped data from a stored XML file
#  * with a timestamp before deciding to pull data again from the Internet.
#  * This method conserves HTTP calls and reduces traffic, allowing for
#  * rapid testing at no risk of disturbing the data source.
#  * 
#  * Requires Python
#  * Assume that if XML file exists, it is formatted correctly.
#  *
#  * Sample zip codes to test with 
#  * 
#  * Princeton : 08540    // for testing normal
#  * Seattle   : 98101    // for testing rain
#  * Miami     : 33101    // for testing heat
#  * Chicago   : 60290    // for testing cool
#  * Caribou, ME : 04736  // for testing cold
#  * Lahaina, HI : 96761  // for testing different time of day
#  * Schenectady, NY : 12345 // for interesting zip code
#  ***************************************************************/
import datetime
import sys
import os.path

# Universal XML parser
def get(xmlText, tag):
	start = xmlText.index("<" + tag + ">") + len(tag) + 2
	end = xmlText.index("</" + tag + ">")
	return xmlText[start:end]

def tempAnalysis(temp, HOT, WARM, COOL, COLD, FREEZING):
    if temp >= HOT:
        print("T-Shirt, Shorts")
    
    elif temp <= HOT and temp >= WARM:
        print("T-Shirt and Shorts + Layer")
    
    elif temp <= WARM and temp >= COOL:
        print("Long Pants, Light Jacket")
    
    elif temp <= COOL and temp >= COLD:
        print("Long Pants, Outer Layer and/or Light Jacket")
    
    elif temp <= COLD and temp >= FREEZING:
        print("Long Pants, Winter Jacket, Hat")
    
    elif temp <= FREEZING:
        print("FREEZING: minimize outdoor exposure")

# Execution

CHRONINTERVAL = 30

HOT = 70
WARM = 60
COOL = 50
COLD = 35
FREEZING = 15

zip = "08540" # default zip code if no argument

if len(sys.argv) > 1:
	zip = sys.argv[1]

# Update subjective weather preferences
if len(sys.argv) > 3:
    COLD = float(sys.argv[2])
    WARM = float(sys.argv[3])
    COOL = (WARM + COLD) / 2
    FREEZING = COLD - 0.5*(COOL - COLD)
    HOT = WARM + 0.5*(WARM - COOL)
    print("Custom Temperature Preferences:")
    print("Hot:     \t" + str(HOT))
    print("Warm:    \t" + str(WARM))
    print("Cool:    \t" + str(COOL))
    print("Cold:    \t" + str(COLD))
    print("Freezing:\t" + str(FREEZING))


filename = "data/" + zip
now = datetime.datetime.now()

if os.path.isfile(filename) == False: 
	os.system("python WeatherScrape.py " + zip)

with open(filename) as f:
    xmlText = f.read()

then = datetime.datetime.strptime(get(xmlText, "time"), "%Y-%m-%dT%H:%M:%S.%f")
thatZip = get(xmlText, "zip")

#  Refresh the weather data if necessary (mismatched location OR data too old)
diff = now - then 
if (thatZip != zip or (diff.seconds / 60) > CHRONINTERVAL):
	os.system("python WeatherScrape.py " + zip)

# Get weather variables
location = get(xmlText, "town") + ", " + get(xmlText, "state")
temp = float(get(xmlText, "temp"))
feel = float(get(xmlText, "feel"))
humidity = float(get(xmlText, "humidity"))
sky = get(xmlText, "sky")

# Handle rain
rainVal = float(get(xmlText, "rain"))
if rainVal == 0:
	rain = False
else:
	rain = True

print("Weather for " + location)
print("Current Temperature is: " + str(temp) + " degrees Fahrenheit")
print("Feels like: " + str(feel) + " degrees Fahrenheit")
print("Humidity: " + str(humidity) + "%")
print("Sky is " + sky)
print("Rain is " + str(rainVal) + " in.")
print("Consider wearing: ")
tempAnalysis(temp, HOT, WARM, COOL, COLD, FREEZING)

if rain:
	print("Bring an umbrella\n")
if sky == "Clear":
	print("Bring sunglasses\n")